### PR TITLE
Add board display for mobile devices

### PR DESCRIPTION
**Pull request by jonathan-marmelab**
This PR closes the following User Story: https://trello.com/c/nyvD7iXR/59-8-as-robert-i-want-to-witness-a-game-on-mobile

- [X] Retrieve state from API
- [X] Display board

Depends on #24

### CODE REVIEW

**Review by djhi — APPROVED** [review]
Well done!

File: mobile/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+install:					## Install dependencies
+	@(if [ -f ${HOME}/.nvm/nvm.sh ] ; \
+	then \
+		. ${HOME}/.nvm/nvm.sh && nvm install 18 && nvm use; \
+	fi)
+
+	@npm install
+
+test: install				## Run Vitest
+
```
Comment (fzaninotto) [inline mobile/Makefile:16]: You shouldn't add newlines in makefile recipes

File: mobile/index.html [hunk 2]
```diff
@@ -5,9 +5,10 @@
     <link rel="icon" type="image/svg+xml" href="/vite.svg" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
     <title>Vite + React + TS</title>
+
+    <script type="module" src="/src/main.tsx" defer></script>
```
Comment (fzaninotto) [inline mobile/index.html:9]: Can you explain the rationale?

File: mobile/src/board/components/BoardView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import PlayerPawnView from "../PlayerPawnView";
+import TileView from "../TileView";
+
+import "./index.css";
+
+interface BoardProps {
+  board: Board;
+}
+
+/**
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:12]: Useless empty comment

File: mobile/src/board/components/BoardView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ */
+const BoardView = ({
+  board: {
+    state: { tiles, remainingTile, gameState, players },
+  },
+}: BoardProps) => {
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:19]: Prefer

`() => (<div>)`

to 

`() => {
  return (<div>)
}`

As it removes one level of indentation and therefore makes the code a bit easier to read.

File: mobile/src/board/components/BoardView/index.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  return (
+    <>
+      <div className="board">
+        {tiles.flatMap((lineTiles, line) =>
+          lineTiles.map((boardTile, row) => (
+            <TileView
+              key={`${line * tiles.length + row}`}
+              boardTile={boardTile}
+              disabled={gameState != GameState.MovePawn}
+              onClick={() => {}}
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:29]: What's that?

File: mobile/src/board/components/BoardView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            >
+              {players
+                .filter(
+                  ({ position }: Player) =>
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:33]: The typehing shouldn't be necessary here

File: mobile/src/board/components/BoardView/index.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                    position.line == line && position.row == row
+                )
+                .map(({ color }: Player) => (
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:36]: neither here

File: mobile/src/board/components/BoardView/index.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            <TileView
+              key={`${line * tiles.length + row}`}
+              boardTile={boardTile}
+              disabled={gameState != GameState.MovePawn}
+              onClick={() => {}}
+            >
+              {players
+                .filter(
+                  ({ position }: Player) =>
+                    position.line == line && position.row == row
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:34]: why not use `===`?

File: mobile/src/board/components/BoardView/index.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                )
+                .map(({ color }: Player) => (
+                  <PlayerPawnView key={`${color}`} color={color} />
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:37]: Prefer

```
<PlayerPawnView key={color} color={color} />
```

to 

```
<PlayerPawnView key={`${color}`} color={color} />
```

File: mobile/src/board/components/BoardView/index.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+                ))}
+            </TileView>
+          ))
+        )}
+      </div>
+      <TileView boardTile={remainingTile} onClick={() => {}} />
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:43]: why the empty onClick?

File: mobile/src/board/components/PlayerPawnView/index.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+import "./index.css";
+
+interface PlayerProps {
+  color: Color;
+}
+
+const PlayerPawnView: FunctionComponent<PlayerProps> = ({
+  color,
+}: PlayerProps): ReactElement => {
```
Comment (fzaninotto) [inline mobile/src/board/components/PlayerPawnView/index.tsx:13]: same, use the `() => JSX` syntax

File: mobile/src/board/components/PlayerPawnView/index.tsx [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+import { Color } from "../../entity";
+
+import "./index.css";
+
+interface PlayerProps {
+  color: Color;
+}
+
+const PlayerPawnView: FunctionComponent<PlayerProps> = ({
```
Comment (fzaninotto) [inline mobile/src/board/components/PlayerPawnView/index.tsx:11]: Don't use the `FunctionComponent` type

File: mobile/src/board/components/TileView/index.tsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+const TileView: FunctionComponent<TileProps> = ({
+  boardTile: {
+    tile: { treasure, shape },
+    rotation,
+  },
+  disabled = false,
+  onClick,
+  children,
+}: TileProps): ReactElement => {
+  return (
```
Comment (fzaninotto) [inline mobile/src/board/components/TileView/index.tsx:56]: same, use `() => JSX`

File: mobile/src/board/components/TileView/index.tsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+interface TileProps {
+  boardTile: BoardTile;
+  disabled?: boolean;
+  onClick: MouseEventHandler;
+  children?: ReactElement | ReactElement[];
+}
+
+const TileView: FunctionComponent<TileProps> = ({
```
Comment (fzaninotto) [inline mobile/src/board/components/TileView/index.tsx:47]: same, don't use `FunctionComponent`

File: mobile/src/board/entity.tsx [hunk 15]
```diff
@@ -0,0 +1,60 @@
+export enum GameState {
```
Comment (fzaninotto) [inline mobile/src/board/entity.tsx:1]: Don't use TypeScript's `enum` type. 

https://www.executeprogram.com/blog/typescript-features-to-avoid

File: mobile/src/board/entity.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  End,
+}
+
+export interface Board {
+  id: number;
+  state: BoardState;
+}
+
+export interface BoardState {
+  tiles: Array<Array<BoardTile>>;
```
Comment (fzaninotto) [inline mobile/src/board/entity.tsx:13]: I find `BoardTile[][]` more legible

File: mobile/src/board/hooks.tsx [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import { useContext, useEffect, useState } from "react";
```
Comment (fzaninotto) [inline mobile/src/board/hooks.tsx:1]: call this file `useBoard` as it only exports that hook. 

In general, we export one hook/component per file.

File: mobile/src/board/hooks.tsx [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+import { Board } from "./entity";
+import { BoardRepository } from "./repository";
+
+export function useBoard(initialState: Board): Board {
+  const [board, setBoard] = useState<Board>(initialState);
+
+  useEffect(() => {
+    const mercureURL = `/.well-known/mercure?topic=${encodeURI(
+      window.location.href
```
Comment (fzaninotto) [inline mobile/src/board/hooks.tsx:11]: I'd say that's fragile... This board can only be used on certain urls, and nothing in its API shows it. 

You'd better pass the topic to the hook, and leave the responsibility of reading it from the window location to the parent. 

In general: move side effects up the chain.

File: mobile/src/board/hooks.tsx [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    )}`;
+
+    const eventSource = new EventSource(mercureURL);
+    eventSource.onmessage = async () => {
+      try {
+        const updatedBoard = await BoardRepository.getById(board.id);
+        setBoard(updatedBoard);
+      } catch (e) {
+        console.error(e);
```
Comment (fzaninotto) [inline mobile/src/board/hooks.tsx:20]: You got an error calling the API to refresh the board, but you say nothing to the user? They'll see an outdated board and do wrong moves. 

You should at least display a snackbar saying that something went wrong and the game isn't up to date.

File: mobile/src/board/hooks.tsx [hunk 20]
```diff
@@ -0,0 +1,30 @@
+import { useContext, useEffect, useState } from "react";
+
+import { Board } from "./entity";
+import { BoardRepository } from "./repository";
+
+export function useBoard(initialState: Board): Board {
```
Comment (fzaninotto) [inline mobile/src/board/hooks.tsx:6]: I find the name a bit... generic.

File: mobile/src/board/pages.tsx [hunk 21]
```diff
@@ -0,0 +1,13 @@
+import { ReactElement, useContext } from "react";
+import { useLoaderData, LoaderFunctionArgs } from "react-router-dom";
+
+import { Board } from "./entity";
+import BoardView from "./components/BoardView";
+import { useBoard } from "./hooks";
+
+export function GetById(): ReactElement {
```
Comment (fzaninotto) [inline mobile/src/board/pages.tsx:8]: That's a bad component name. Do you mean `<Board>`?

File: mobile/src/board/pages.tsx [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+  const initialState = useLoaderData() as Board;
+  const board = useBoard(initialState);
```
Comment (fzaninotto) [inline mobile/src/board/pages.tsx:10]: This isn't very readable. I understand that the hook subscribes to the board updates. Why not write it as:

```jsx
const initialBoard = useLoaderData<Board>();
const board = useUpdateBoardOnRemoteChange(initialBoard);
```

File: mobile/src/board/repository.tsx [hunk 23]
```diff
@@ -0,0 +1,11 @@
+import { Board } from "./entity";
+
+export class BoardRepository {
```
Comment (fzaninotto) [inline mobile/src/board/repository.tsx:3]: A static class with stateless functions is unusual in React, where we favor functional programming. Js already has the notion of modules, so you don't need static functions. I'd just export the `getById` function.

File: mobile/src/index.css [hunk 24]
```diff
@@ -1,69 +1,127 @@
...
+  width: 100%;
+  height: 100%;
+  gap: 20px;
+  align-items: center;
+  justify-content: flex-end;
+}
+
+header form input[type="text"] {
+  height: 100%;
+  background-color: #444444;
```
Comment (fzaninotto) [inline mobile/src/index.css:93]: #444 is shorter ;)

File: mobile/src/shared/components/Layout.tsx [hunk 25]
```diff
@@ -0,0 +1,15 @@
+import type { FunctionComponent, ReactElement } from "react";
+import { Outlet } from "react-router-dom";
+
+const Layout: FunctionComponent<{}> = (): ReactElement => {
+  return (
```
Comment (fzaninotto) [inline mobile/src/shared/components/Layout.tsx:5]: `() => JSX`

File: mobile/src/shared/components/Layout.tsx [hunk 26]
```diff
@@ -0,0 +1,15 @@
+import type { FunctionComponent, ReactElement } from "react";
+import { Outlet } from "react-router-dom";
+
+const Layout: FunctionComponent<{}> = (): ReactElement => {
```
Comment (fzaninotto) [inline mobile/src/shared/components/Layout.tsx:4]: No use for `FunctionComponent` or `ReactElement`, TypeScript infers it well

File: mobile/src/board/components/BoardView/index.css [hunk 27]
```diff
@@ -0,0 +1,16 @@
+:root {
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.css:1]: In general, we avoid arranging files by type (component, hook) and using that type to name folders. Read https://marmelab.com/blog/2015/12/17/react-directory-structure.html for the rationale.

File: mobile/src/board/components/BoardView/index.tsx [hunk 28]
```diff
@@ -0,0 +1,48 @@
+import { Board, Player, GameState } from "../../entity";
```
Comment (fzaninotto) [inline mobile/src/board/components/BoardView/index.tsx:1]: I'm not a fan of having a gazillion files named "index.tsx" - it really becomes a pain to find/navigate between files in the IDE

File: mobile/src/board/entity.tsx (hunk 15 above)
Comment (fzaninotto) [inline mobile/src/board/entity.tsx:1]: This file only contains types. The convention is to call it `types.ts`(no need for `.tsx` as there is no JSX)

**Review by fzaninotto — COMMENTED** [review]
I'm a bit late, but I have a few remarks...

**Pull request merged**